### PR TITLE
Sk/test refactor

### PR DESCRIPTION
**Pull request by snopoke**
Easiest reviewed by commit

## Technical Summary
Our current controller tests mock the controller manually which does work but has some limitations (e.g. properties don't work https://github.com/dimagi/formplayer/pull/1209#discussion_r965834890).

See https://github.com/dimagi/formplayer/pull/1227/files#diff-a037efdee62087635e2f354d4f79a00fa76ec31beaabc7a2bcdbbbfe2aad389d for a test that actually uses this.

This PR makes a start to bring out tests inline with Spring testing patterns.

The primary change is to stop mocking the controllers and allow Spring to set them up:

```
@WebMvcTest
@ContextConfiguration(classes = [TestContext::class, CacheConfiguration::class])
@Import(FormController::class, FormSubmissionController::class)
class MediaCaptureTest {
```

This test uses the `@Import` annotation to bring the controllers under test into the web app context and have them configured by Sprint (technically you should be able to pass the controllers directly to the `@WebMvcTest` annotation but for some reason that isn't working).

The dependencies for the controllers (e.g. FormSessionService) are provided by the context classes passed to `@ContextConfiguration`.

With this setup all that's needed is configuring other mocks like `RestoreFactory`.

**Configuring mocks**
To handle this I've tried a different approach rather than having a very large base class that does all the setup. I've split the mock setup into separate classes for each mock and made them into Junit5 extensions which can be used to annotate the test class.

This allows us to keep the mock setup compartmentalised and also makes it possible to use only those that are needed for a test.

For convenience we can bundle a number of extension together using an annotation which I've done here (see `FormSessionTest`).

I have not attempted to change any other tests but I have replaced relevant parts of the `BaseTestClass` with the new mock setup classes to reduce code duplication.

**Making requests**
The other thing that I've done is to extract the code which makes the mock MVC requests into their own classes instead of using the generic methods in the base class which are complex and hard to follow. See the `org.commcare.formplayer.junit.request` package.

### Safety story
Only affects tests

### QA Plan
If the build passes

### Special deploy instructions
- [x] This PR can be deployed after merge with no further considerations.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations.

### Review

- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change.

### CODE REVIEW

**Review by shubham1g5 — CHANGES_REQUESTED** [review]
Left some minor comments though this looks so much better than earlier!

File: src/test/java/org/commcare/formplayer/junit/MockTimezoneProvider.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package org.commcare.formplayer.junit;
+
+import org.javarosa.core.model.utils.TimezoneProvider;
+
+public class MockTimezoneProvider extends TimezoneProvider {
```
[the remark this answers is not in the document]
Comment (snopoke) [reply src/test/java/org/commcare/formplayer/junit/MockTimezoneProvider.java:5]: It was already duplicated in `BaseTestClass`. This was meant to move it from there but I forgot the delete: https://github.com/dimagi/formplayer/pull/1228/commits/3c491012e1c2327188f1f97bdfb884791c4c30ba

We can't currently use the one in core without adding a dependency on the `commcare.test` module.

File: src/test/java/org/commcare/formplayer/junit/InitializeStaticsExtension.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    public void beforeEach(ExtensionContext context) throws Exception {
+        LocalizerManager.setUseThreadLocalStrategy(true);
+        Localization.getGlobalLocalizerAdvanced().addAvailableLocale("default");
+        Localization.setLocale("default");
+        ReferenceManager.instance().addReferenceFactory(new ClasspathFileRoot());
+        Localization.registerLanguageReference("default",
+                "jr://springfile/formplayer_translatable_strings.txt");
+
+        PrototypeUtils.setupThreadLocalPrototypes();
+        new SQLiteProperties().setDataDir("testdbs/");
```
[the remark this answers is not in the document]
Comment (snopoke) [reply src/test/java/org/commcare/formplayer/junit/InitializeStaticsExtension.java]: I see that `SQLiteProperties` already uses the `sqlite.dataDir` property so we can just use that. I've pushed some more commits that updates the `GenerateSnapshotDatabases` test.

Incidentally, where does that test get executed. In normal test runs it's a noop but there is a gradle task to call it: https://github.com/dimagi/formplayer/blob/40110bf84bbf59af4816c03dbca5f4d82d2d77ca/build.gradle#L165-L172

But I don't know where that's called.

[the remark this answers is not in the document]
Comment (snopoke) [reply src/test/java/org/commcare/formplayer/junit/InitializeStaticsExtension.java]: I see it is referenced here: https://github.com/dimagi/formplayer/blob/73ef9df1261fb86502788b6cdb94f5bd7df2cf83/src/test/java/org/commcare/formplayer/tests/SnapshotTests.java#L93-L95

File: src/test/java/org/commcare/formplayer/utils/GenerateSnapshotDatabases.java [hunk 3]
```diff
@@ -63,8 +83,13 @@ public void testCreateSnapshot() throws Exception {
             if (new File(SQLiteProperties.getDataDir()).exists()) {
                 fail("Couldn't remove existing snapshot assets");
             }
-            syncDb();
-            doInstall("sandbox_reference/snapshot.json");
+            new SyncDbRequest(mockMvc, restoreFactory).request();
```
Comment (shubham1g5) [inline src/test/java/org/commcare/formplayer/utils/GenerateSnapshotDatabases.java:86]: we should probably delete `syncDb()` from `BaseTestClass` and replace it with the new `SyncDbRequest`

File: src/test/java/org/commcare/formplayer/tests/BaseTestClass.java [hunk 4]
```diff
@@ -286,127 +268,18 @@ public void setUp() throws Exception {
...
 
-        doAnswer(new Answer<Void>() {
-            @Override
-            public Void answer(InvocationOnMock invocation) throws Throwable {
-                String key = (String)invocation.getArguments()[0];
-                sessionMap.remove(key);
-                return null;
-            }
-        }).when(formSessionService).deleteSessionById(anyString());
+        new SQLiteProperties().setDataDir("testdbs/");
```
Comment (shubham1g5) [inline src/test/java/org/commcare/formplayer/tests/BaseTestClass.java:275]: Do we need it here as well when we are doing the same in `TestContext` ?

**Review by shubham1g5 — COMMENTED** [review]
Minor comments,  a stickler pass would be great as well.

File: src/test/java/org/commcare/formplayer/tests/BaseTestClass.java (hunk 4 above)
Comment (snopoke) [reply src/test/java/org/commcare/formplayer/tests/BaseTestClass.java:275]: no I don't think so

Comment (snopoke) [reply src/test/java/org/commcare/formplayer/tests/BaseTestClass.java:277]: turns out it is needed - I'm not sure why

File: src/test/java/org/commcare/formplayer/junit/InitializeStaticsExtension.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply src/test/java/org/commcare/formplayer/junit/InitializeStaticsExtension.java]: Yeah it's supposed to run manually to update the db snapshots in case of app db changes. There is more info in Javadoc for the `GenerateSnapshotDatabases` class.

**Review by shubham1g5 — APPROVED** [review]
